### PR TITLE
Humanoid appearance refactor

### PR DESCRIPTION
**Pull request by vulppine**
# Humanoid appearance refactor

Refactors the way that human(oid) appearances are handled by SS14 client and server side,
as well as how they're constructed in prototypes - both through entities, marking accessories,
and species-specific sprites.

This currently combines four different systems, managers, and prototypes under the new Humanoid namespace:
- CharacterAppearance/HumanoidAppearance
- SpriteAccessory
- Markings
- Species

No prototype definitions that already exist were deprecated, except for SpriteAccessory, 
which was merged into Markings. There is a tool to assist in migrating any old SpriteAccessories
to Markings, which can be found in `Tools/markings`. This requires the Go toolchain
to use.

## How to use

The refactor changes several things. 

**Humanoid entities**

- Anything that derives from the base human
  no longer needs to define their own sprite component. This can be removed entirely, unless
  you have specific needs for SpriteComponent - instead, you don't have to define the
  specific layers. You still can define your own version of sprite component if you want to,
  however.
- Anything that derives from the base human also no longer has to have a 'Markings' component.
  This can be removed entirely.
- All humanoids should now have the 'Humanoid' component, with the minimum of the 'species'
  field filled in. This allows for that humanoid to be spawned in-game with the correct
  appearance settings (e.g., a reptilian has a Humanoid component with `species: Reptilian`)
- Humanoids with RandomHumanoidAppearance are now more randomized than before, and can be different
  species as well. Expect to see more lizards and slimes where you wouldn't see lizards and slimes.
- Humanoids can also have a pre-determined appearance/base layer set, which can be shared across entities. Just define
  a prototype of type `HumanoidProfilePrototype`, and give its ID to a humanoid component under the field 'initial'.

**Humanoid markings/SpriteAccessories**

- All sprite accessories were moved to the Marking system, which allows for flexibility
  when it comes to restricting species from having specific types of markings without having
  to hard code values (no more HumanHair and VoxHair enum values), as well as the ability
  to have multiple sprites on a single accessory.
- This may result in the Hair/FacialHair columns in DB to be deprecated sometime in the future.

**Humanoid species**

- Humanoid species now have two new fields for construction: `sprites` and `markingLimits`.
- `sprites` defines the 'base sprites' of an entity, which is an abstraction over the aforementioned
  SpriteComponent that is no longer required on a humanoid entity. This also defines how markings
  are displayed on a humanoid entity, and deals with things such as matching skin tone, or allowing
  markings to appear on the entity.
- `markingLimits` is functionally equivalent to the original Marking's `markingPoints` field,
  but you can now specify if a species is only allowed to have whitelisted markings or not
  (for examples, see Reptilian and Vox)

**Profile editor**

- The hair/facial hair setters have been changed to match the new usage of markings as the backing
  system for hair. You can now 'remove' hair (which auto sets it to bald).
- The hair/facial hair setters also disappear if a species does not have any hair to use. For example,
  reptilians, which were never able to use hair, will never see the hair picker in the first place.
- The random button is now more randomized, similar to humanoid entities, and can now roll species
  with random skin tones. No more human-colored slimes or lizards in randomization.
- Humanoid profiles now have more verification, and skin tones are now verified with the other properties
  of a humanoid profile.
- Administrators with the +FUN flag can now set appearances in-round using a new tool located in tricks.
  This allows an administrator to set the base sprites of an entity, as well as their color, and equally
  add as many markings to a player character as they want. (i.e., you can now force people to have cat ears
  and lizard tails)

## Why

Humanoid appearance was scattered through four different systems, and the API for interacting with
humanoid appearance was also not very fun to use - this attempts to make it so that the API is nicer
to work with in an ECS context, and it also makes it so that humanoids and species are more cleaner
to construct in terms of prototypes.

## The future

- Vox, which will still be disabled round-start, are completely constructable in this refactor,
  and require item sprite changes in order to be round-start ready.
- Body system still requires a refactor - once this is reimplemented, humanoid appearance will
  inevitably need to be revisited. I did this in hopes of at least taking humanoid appearance
  a step in a good direction, so that whoever inevitably has to do body system (hopefully) won't have to
  deal too much with the idea of humanoid character appearances.

TODO:
- [x] Magic Mirror
- [x] File organization
- [x] Randomization
- [x] ~~Commands~~ In-round modification window (close enough)

:cl:

- tweak: Humanoids were recently changed. Expect to see more variations of humanoids in places
  that may have been primarily dominated by humans.

### CODE REVIEW

File: Resources/Prototypes/Species/vox.yml [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    Chest: MobVoxTorso
+    Eyes: MobVoxEyes
+    LArm: MobVoxLArm
+    RArm: MobVoxRArm
+    LHand: MobVoxLHand
+    RHand: MobVoxRHand
+    LLeg: MobVoxLLeg
+    RLeg: MobVoxRLeg
+    LFoot: MobVoxLFoot
+    RFoot: MobVoxRFoot
```
Comment (moonheart08) [inline Resources/Prototypes/Species/vox.yml:26]: Wait so how would all this work with, say, surgery, if you swapped someone's limbs out?

Comment (vulppine) [reply Resources/Prototypes/Species/vox.yml:26]: you would have to do a call that adds a 'custom base sprite' to a humanoid component - so if you wanted to swap somebody's limb with a lizard limb or a slimeperson limb, you'd just add the base sprite to the person, and the client would deal with it

File: Content.Client/Clothing/ClothingVisualsSystem.cs [hunk 2]
```diff
@@ -150,6 +150,7 @@ private void OnVisualsChanged(EntityUid uid, ClientInventoryComponent component,
 
     private void OnGotUnequipped(EntityUid uid, ClothingComponent component, GotUnequippedEvent args)
     {
+        /*
         if (component.InSlot == "head"
             && _tagSystem.HasTag(uid, "HidesHair")
             && TryComp(args.Equipee, out SpriteComponent? sprite))
```
Comment (AJCM-git) [inline Content.Client/Clothing/ClothingVisualsSystem.cs:156]: leftover?

File: Content.Client/Clothing/ClothingVisualsSystem.cs [hunk 3]
```diff
@@ -160,6 +161,7 @@ private void OnGotUnequipped(EntityUid uid, ClothingComponent component, GotUneq
             if (sprite.LayerMapTryGet(HumanoidVisualLayers.Hair, out var hair))
                 sprite[hair].Visible = true;
         }
+        */
```
Comment (AJCM-git) [inline Content.Client/Clothing/ClothingVisualsSystem.cs:164]: leftover?

File: Content.Client/Clothing/ClothingVisualsSystem.cs [hunk 4]
```diff
@@ -198,6 +200,7 @@ private void OnGotEquipped(EntityUid uid, ClothingComponent component, GotEquipp
     {
         component.InSlot = args.Slot;
 
+        /*
         if (args.Slot == "head"
             && _tagSystem.HasTag(uid, "HidesHair")
             && TryComp(args.Equipee, out SpriteComponent? sprite))
```
Comment (AJCM-git) [inline Content.Client/Clothing/ClothingVisualsSystem.cs:206]: leftover?

File: Content.Client/Clothing/ClothingVisualsSystem.cs [hunk 5]
```diff
@@ -208,6 +211,7 @@ private void OnGotEquipped(EntityUid uid, ClothingComponent component, GotEquipp
             if (sprite.LayerMapTryGet(HumanoidVisualLayers.Hair, out var hair))
                 sprite[hair].Visible = false;
         }
+        */
```
Comment (AJCM-git) [inline Content.Client/Clothing/ClothingVisualsSystem.cs:214]: leftover?

**Review by AJCM-git — CHANGES_REQUESTED** [review]
Just some nitpicks about unused code. I also noticed that you dont leave a line between markings prototypes which makes them look very cramped

**Comment by vulppine** [comment]
> Just some nitpicks about unused code. I also noticed that you dont leave a line between markings prototypes which makes them look very cramped

if you're talking about the hair prototypes, that's just a result of the conversion tool

**Review by moonheart08 — APPROVED** [review]

**Comment by Elijahrane** [comment]
random humanoid appearance probably needs some way to blacklist certain species, even being playable doesn't guarantee it's fine to become a nukie or whatever

File: Content.Server/Humanoid/Components/RandomHumanoidAppearanceComponent.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Content.Server.CharacterAppearance.Components;
+
+[RegisterComponent]
+public sealed class RandomHumanoidAppearanceComponent : Component
+{
+    [DataField("randomizeName")] public bool RandomizeName = true;
+
+    [DataField("ignoredSpecies")] public readonly HashSet<string> IgnoredSpecies = new();
```
Comment (metalgearsloth) [inline Content.Server/Humanoid/Components/RandomHumanoidAppearanceComponent.cs:8]: customtypeserializer?

**Comment by moonheart08** [comment]
hmm written by a maint and two approv

what if i just

**Pull request merged**